ts: Migrate `list_widget` to TypeScript.

Lalit3716 opened the pull request:
Migrates the `list_widget` to TypeScript.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Lalit3716 commented:
@zulipbot add "mentor review"

PIG208 commented:
@zulipbot add "area: typescript migration"

PIG208 commented on web/src/list_widget.js:
```diff
@@ -284,15 +295,16 @@ export function create($container, list, opts) {
         // the sorting function is either the function or string that calls the
         // function to sort the list by. The prop is used for generic functions
         // that can be called to sort with a particular prop.
```
This comment needs to be updated. Because we have removed `prop`, and that we expect a different format for `sorting_function` if it is a `string`.

Lalit3716 replied:
Oh yeah, nice catch!

Lalit3716 commented on web/src/list_widget.ts:
```diff
@@ -163,7 +189,7 @@ export function create($container, list, opts) {
         offset: 0,
         list,
-        filtered_list: list,
+        filtered_list: [],
```
Yeah, we can be sure that `filtered_list` will have valid data before we access it via `get_current_list`, because we call `widget.clean_redraw` when we create a list widget which sets the correct data for `filtered_list`.
https://github.com/zulip/zulip/blob/ba54dfd83cc05f8c9fb0f602bb3d2576f35d6ef5/web/src/list_widget.ts#L476

PIG208 commented on web/src/list_widget.ts:
```diff
@@ -3,6 +3,28 @@ import $ from "jquery";
+    clear_event_handlers(): void;
+    increase_rendered_offset(): void;
+    reduce_rendered_offset(): void;
+    remove_rendered_row(row: HTMLElement): void;
```
I think we are getting a JQuery element from the call site:
https://github.com/zulip/zulip/blob/27878cce87c973100979e8a6fd8b33705b934fee/web/src/recent_topics_ui.js#L489-L493
https://github.com/zulip/zulip/blob/27878cce87c973100979e8a6fd8b33705b934fee/web/src/recent_topics_ui.js#L586-L608

PIG208 replied:
`topic_row` is not consistent with our naming scheme for JQuery elements with a `$` prefix, so perhaps that can be fixed too.

PIG208 commented on web/src/list_widget.ts:
```diff
@@ -17,6 +17,31 @@ type ListWidgetMeta<Key = unknown, Item = Key> = {
+    modifier: (item: Item) => string;
+    init_sort?: string | SortingFunction<Item>;
+    initially_descending_sort?: boolean;
+    html_selector?: (item: Item) => JQuery;
```
In addition to `(item: Item) => Jquery`, `(item: Item) => string` is also acceptable.

Lalit3716 replied:
I took a look and it seems there was only one place where `html_selector` was returning string. So I think it makes sense to return a `JQuery` collection there instead of string and avoid writing type like `string | JQuery` ?
This is the one place where we are returning string(settings_users.js) - 
https://github.com/zulip/zulip/blob/784622ee5d3e5bca4db9ebafba748acd797b427d/web/src/settings_users.js#L290

PIG208 replied:
Yeah, also it seems that `JQuery` instead of `string` is expected from `html_selector`:
https://github.com/zulip/zulip/blob/d6c21c2237918f402988bd7e9463afa9253a1192/web/src/list_widget.js#L395-L401
I'm not sure if this works previously for `settings_users` and whether it was an intentional decision to use a `string` there.

Lalit3716 replied:
Yeah I think that `insert_rendered_row` was written only for use in `recent_topics_ui` and I don't think that this method previously worked for `settings_users`.

PIG208 commented on web/src/list_widget.ts:
```diff
@@ -17,6 +17,31 @@ type ListWidgetMeta<Key = unknown, Item = Key> = {
+    filter?: {
+        $element: JQuery;
+        predicate?: (item: Item, value: string) => boolean;
+        filterer?: (list: Item[], value: string) => Item[];
```
We can type `predicate` and `filterer` as being mutually exclusive.

Lalit3716 replied:
Oh right we should do that 👍

PIG208 commented on web/tests/list_widget.test.js:
```diff
@@ -561,66 +561,66 @@ run_test("clear_event_handlers", () => {
+//         get_item: (item) => item,
+//     });
+//     blueslip.reset();
+// });
```
Should we just remove the test cases?

Lalit3716 replied:
Yeah I think it makes sense to remove them now that we don't have over-defensive code to raise those errors 👍

PIG208 commented:
Left some more comments. Thanks for putting this all together! @Lalit3716

Lalit3716 commented:
@PIG208 Thanks for the review!
Updates:
- Made `predicate` and `filterer` options mutually exclusive.
- Refined some types.

This is ready for another look!

PIG208 commented on web/src/list_widget.ts:
```diff
@@ -540,11 +548,11 @@ export function create<Key = unknown, Item = Key>(
-export function get(name) {
-    return DEFAULTS.instances.get(name) || false;
+export function get(name: string): ListWidget | false {
+    return DEFAULTS.instances.get(name) ?? false;
```
Does this need to return `false` as a fallback? Might be worth a follow-up PR as a quick refactoring project.

Lalit3716 replied:
Yeah, I am planning to refactor this module a bit after it gets translated to TypeScript so that I will know I am not breaking
things when doing refactoring due to all those nice typescript errors. So yeah I will do a follow-up refactoring project :)

PIG208 commented on web/tests/list_widget.test.js:
```diff
@@ -756,7 +756,7 @@ run_test("render item", () => {
         modifier: (item) => `<tr data-item=${item.value}>${item.text}</tr>\n`,
         get_item,
-        html_selector: (item) => `tr[data-item='${item}']`,
+        html_selector: (item) => `tr[data-item='${item.value}']`,
```
`html_selector` should return `JQuery`. Same thing for `widget_3` in this test case.

PIG208 commented on web/src/list_widget.ts:
```diff
@@ -468,9 +474,11 @@ export function create<Key = unknown, Item = Key>(
             }
-            if (!opts.filter.predicate(item)) {
+
+            if (!opts.filter?.predicate?.(item, meta.filter_value)) {
```
I think we want `!.` or assertions here. `?.` makes it silently return when either `filter` or `predicate` is missing. It's better to fail here to indicate that there is a programming error.

PIG208 commented on web/src/list_widget.ts:
```diff
@@ -231,10 +190,6 @@ export function create<Key = unknown, Item = Key>(
         return undefined;
```
This can be removed as well. `!opts` is always falsy.

PIG208 commented:
https://github.com/zulip/zulip/blob/5734b9abc4064b38073c301fdccd07c3c89b117a/web/src/list_widget.ts#L436

The first condition is redundant.

PIG208 commented:
This seems like a bug unrelated to the PR:
https://github.com/zulip/zulip/blob/5734b9abc4064b38073c301fdccd07c3c89b117a/web/src/list_widget.ts#L293-L298
`$.find` always returns a truthy value, so the check that follows is redundant.

PIG208 commented:
I went over the PR again and posted some comments.
The prep commit sequence (first 4 commits) should be ready for an integration review.
The rest of them have some minor issues to be addressed. I ran eslint with a stricter setting to double-check. Eventually, we should have more confident assurance of its correctness as we progress further into the migration.

Lalit3716 commented:
Updates:
- Addressed reviews left by @PIG208 

Question:
Regarding this comment [comment](https://github.com/zulip/zulip/pull/25326#issuecomment-1614096565) I think we want to check the length of the JQuery collection here instead of checking the truthiness of the JQuery collection itself ?

PIG208 commented:
> Question: Regarding this [comment](https://github.com/zulip/zulip/pull/25326#issuecomment-1614096565) I think we want to check the length of the JQuery collection here instead of checking the truthiness of the JQuery collection itself ?

Yeah, that sounds correct to me.

PIG208 commented on web/src/list_widget.ts:
```diff
@@ -346,7 +287,7 @@ export function create<Key = unknown, Item = Key>(
             }
             const $html_item = meta.$scroll_container.find(opts.html_selector(item));
-            if (!$html_item) {
+            if ($html_item.length === 0) {
```
I think this fix goes to a separate commit.

Lalit3716 commented:
@PIG208 Made a new commit for the suggested change, ready for another look :)

PIG208 commented:
Looks good to me!

timabbott commented on web/src/list_widget.js:
```diff
@@ -495,3 +487,5 @@ export function handle_sort($th, list) {
     list.sort(sort_type, prop_name);
 }
+
+export const default_get_item = (item) => item;
```
I can merge this as written, but I'm not convinced we wouldn't be happier with the default being set inside the `ListWidget` constructor if not passed; that seems a lot cleaner than making 20 files all ask for the identity function...

Lalit3716 replied:
Definitely, it's just a hack for now based on the conversation we had here - https://chat.zulip.org/#narrow/stream/6-frontend/topic/list_widget.20typescript.20migration/near/1558934.

But the future plan here is that all these list_widget instances that use the default identity function should instead be using ids so that they will be passing actual meaningful `get_item` callbacks.

timabbott replied:
OK

timabbott commented on web/src/recent_topics_ui.js:
```diff
@@ -859,6 +859,7 @@ export function complete_rerender() {
         sort_fields: {
             stream_sort,
             topic_sort,
+            ...ListWidget.generic_sort_functions("numeric", ["last_msg_id"]),
```
How can I verify that the diffs in this commit don't change the sorting configuration?

Lalit3716 replied:
I believe for that you will require some manual work 🥲, one thing I already did to verify that the sorting configuration remains the same was to check the corresponding templates which have these table headers which sets the `data-sort` prop and manually see which of these are using generic props.

timabbott replied:
OK, I'll trust you to have checked it carefully; it certainly seems plausible that this configuration is the same as before.

timabbott commented on web/src/list_widget.ts:
```diff
@@ -184,14 +199,15 @@ export function create<Key = unknown, Item = Key>(
         filtered_list: [],
         reverse_mode: false,
         filter_value: "",
+        $scroll_container: scroll_util.get_scroll_element(opts.$simplebar_container),
```
Where does this line come from?

Lalit3716 replied:
This was coming from `set_up_event_handlers` function well I actually extracted this change in a separate commit now with a clear message describing this change.

timabbott commented on web/src/settings_users.js:
```diff
@@ -380,7 +380,7 @@ export function update_bot_data(bot_user_id) {
     }
 
-    bot_list_widget.render_item(bot_user_id);
+    bot_list_widget.render_item(bot_info(bot_user_id));
```
What's the story with this change? It seems like a functional change unrelated to the typescript adjustment.

timabbott commented on web/src/list_widget.ts:
```diff
@@ -428,7 +435,7 @@ export function create<Key = unknown, Item = Key>(
-            if (opts.filter && opts.filter.$element) {
+            if (opts.filter?.$element) {
                 opts.filter.$element.off("input.list_widget_filter");
             }
```
Should the if statement here be removed in favor of the next line using the `?`?

timabbott commented:
Posted a handful of comments, otherwise lgtm.

Lalit3716 commented:
@timabbott Addressed your review and left some comments for you. Also I have extracted two more commits for some changes which you pointed out in your review :)

timabbott commented:
This looks great, squashed the commits marked for squashing and merged, thanks for all the hard work on this @Lalit3716!
